codex/manifest/manifest.nim

```diff
@@ -153,6 +153,37 @@ proc cid*(self: Manifest): ?!Cid =
 # Constructors
 ############################################################
 
+func copyAllScalarFields(
```

Comment:
I'm not entirely sure what the benefit of doing this is, looks like a non-standard way of initializing an object - this isn't very nim'ish...

I think a better approach would be to instead call the default `Manifest.new()` from the protected `new` and then set the remaining protected fields.

Reply:
I don't understand your idea pleas give me a sketch.

We have two operations, protect and unprotect, that essentially copy contructors. They copy most fields and change a few. But I can't use things like shallowCopy because
- copying entire `blocks` arrays is inefficient sincew we anyway will replace it immediately
- we shouldn't copy `protected` field because we need to change it, and Nim doesn't allow changing discriminator fields

so, both `protect` and `unprotect` has to copy all fields except `blocks` and give new value to `protected` at the costruction time. I just factotred out the common code in the single procedure. And I don't like even this solution because new fields added to Manifest will not appear here magically

Reply:
I was talking about doing something like:

```nim
func protect*(self: Manifest, K, M: int) =
  var protected = Manifest.new(self.rootHash,
    self.originalBytes,
    self.blockSize,
    @[],
    self.version,
    self.hcodec,
    self.codec)

  var protected.protected = true
  protected.originalCid = self.cid
  protected.originalLen = self.len
  protected.K = K
  protected.M = M

  ...
  protected
```

Reply:
current code avoids duplication of common code (Manifest.new with the long field list) in `protect` and `unprotect` by moving it into copyAllScalarFields. So you propose to remove copyAllScalarFields and put this code directly into `protect` and `unprotect`? I tried to avoid code duplication, especially since it depends on the field list, so when new fields are added, one will have to modify both functions.